src/modules/rlm_smtp/rlm_smtp.c

```diff
@@ -0,0 +1,897 @@
+	char const		*bcc_addrs;		//!< Comma separated list of emails not to be listed
+	fr_time_delta_t 	timeout;		//!< Timeout for connection and server response
+	fr_curl_tls_t		tls;			//!< Used for handled all tls specific curl components
+	char const		*name;			//!< Auth-Type value for this module instance.
+	fr_dict_enum_t		*auth_type;
+	vp_map_t		*header_maps;			//!< Attribute map used to process header elements
+	CONF_SECTION		*cs;
+	bool 			set_date;
+} rlm_smtp_t;
+
+typedef struct {
+	rlm_smtp_t const    	*inst;		//!< Instance of rlm_smtp.
+	fr_curl_handle_t    	*mhandle;	//!< Thread specific multi handle.  Serves as the dispatch and coralling structure for smtp requests
+} rlm_smtp_thread_t;
+
+/*
+ *	Holds the context for parsing the email elements
+ */
+typedef struct {
+	REQUEST			*request;
+	fr_curl_io_request_t	*randle;
+	fr_cursor_t		cursor;
+	fr_cursor_t		body_cursor;
+	fr_dbuff_t		vp_in;
+	struct curl_slist	*recipients;
+	struct curl_slist	*header;
+	struct curl_slist 	*body_header;
+	fr_time_t 		time;
+	char 			time_str[60];
+	curl_mime		*mime;
+} fr_mail_ctx;
+
+/*
+ *	A mapping of configuration file names to internal variables.
+ */
+static const CONF_PARSER module_config[] = {
+	{ FR_CONF_OFFSET("uri", FR_TYPE_STRING, rlm_smtp_t, uri) },
+	{ FR_CONF_OFFSET("template_directory", FR_TYPE_STRING, rlm_smtp_t, template_dir) },
+	{ FR_CONF_OFFSET("sender_address", FR_TYPE_STRING, rlm_smtp_t, sender_address) },
+	{ FR_CONF_OFFSET("envelope_address", FR_TYPE_STRING, rlm_smtp_t, envelope_address) },
+	{ FR_CONF_OFFSET("recipients", FR_TYPE_STRING, rlm_smtp_t, recipient_addrs) },
+	{ FR_CONF_OFFSET("TO", FR_TYPE_STRING, rlm_smtp_t, to_addrs) },
+	{ FR_CONF_OFFSET("CC", FR_TYPE_STRING, rlm_smtp_t, cc_addrs) },
+	{ FR_CONF_OFFSET("BCC", FR_TYPE_STRING, rlm_smtp_t, bcc_addrs) },
+	{ FR_CONF_OFFSET("timeout",FR_TYPE_TIME_DELTA, rlm_smtp_t, timeout) },
+	{ FR_CONF_OFFSET("set_date", FR_TYPE_BOOL, rlm_smtp_t, set_date), .dflt = "yes" },
+	{ FR_CONF_OFFSET("tls", FR_TYPE_SUBSECTION, rlm_smtp_t, tls), .subcs = (void const *) fr_curl_tls_config },//!<loading the tls values
+	CONF_PARSER_TERMINATOR
+};
+/*
+ * 	Returns the proper envolope address
+ */
+static const char * get_envelope_address(fr_mail_ctx *uctx, rlm_smtp_t const *inst)
+{
+	VALUE_PAIR 	*vp = NULL;
+
+	/* Check if a sender address has been provided */
+	vp = fr_cursor_iter_by_da_init(&uctx->cursor, &uctx->request->packet->vps, attr_smtp_sender_address);
+
+	/* If there was a sender address in the request, use it as the envelope address */
+	if(vp) return vp->vp_strvalue;
+
+	/* If there is an envelope address in the config, use that, otherwise use sender_address */
+	return (inst->envelope_address ? inst->envelope_address : inst->sender_address);
+}
+
+/*
+ * 	Adds every element associated with a dict_attr to a curl_slist
+ */
+static int da_to_slist(fr_mail_ctx *uctx, struct curl_slist **out, const fr_dict_attr_t *dict_attr)
```

Comment:
Good abstraction 👍

Reply:
Hmm, but the problem is this should really be taking an array of `vp_tmpl_t`, so it's more `tmpl_to_slist`